src/client/Microsoft.Identity.Client/ClientApplicationBase.cs

```diff
@@ -426,6 +426,21 @@ public AcquireTokenSilentParameterBuilder AcquireTokenSilent(IEnumerable<string>
                 loginHint);
         }
 
+        /// <summary>
+        /// Retrieve the TenantId for an Authority URL.
+        /// </summary>
+        /// <param name="authorityUri">The Authority URL to parse.</param>
+        /// <returns>The Tenant Id</returns>
+        /// <remarks>
+        /// The Tenant Id can be NULL if the Authority Type is ADFS
+        /// </remarks>
+        public static string GetTenantId(Uri authorityUri)
```

Comment:
@bgavrilMS The developer experience would be something like this:
```csharp
IConfidentialClientApplication cca = ConfidentialClientApplicationBuilder
    .Create(clientId)
    .WithAuthority(authority)
    .WithCertificate(certificate)
    .Build();

AuthenticationResult result = await cca.AcquireTokenForClient(scope)
    .WithTenantId(ClientApplicationBase.GetTenantId(new Uri(authority));)
    .ExecuteAsync()
    .ConfigureAwait(false);
```

Having this method on ClientAppBase seems unintuitive since we don't use the base class for anything. Wondering if there's a better place to put this method and it would be more discoverable (we can document it but still). Can't really add it to ICCA or IPCA... Maybe we could add an overload `WithTenantId` that accepts authority URI and extracts tenant from it?

Reply:
In fact, `WithTenantId` creates an authority and assigns it to AuthorityOverride. WithAuthority does the same thing. For the ask in the issue, maybe the suggestion should be to use WithAuthority.

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/493d76238f6da3d84eea07706704ec4cd33d5d0e/src/client/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs#L277-L297
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/493d76238f6da3d84eea07706704ec4cd33d5d0e/src/client/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs#L136-L143

Reply:
This is a good point, @pmaytak, `.WithTenantFromAuthority(authority)` or something like that would be more discoverable!

```CSharp
AuthenticationResult result = await cca.AcquireTokenForClient(scope)
    .WithTenantFromAuthority(authority)
    .ExecuteAsync()
    .ConfigureAwait(false);
```

Reply:
Hmm...   It does look like `.WithAuthority(authority)` already does everything that is needed...

Reply:
The problem with `WithAuthority` is that there are authority aliases which don't always work, e.g. I start with `login.microsoft.com/common` and I read the authority from a WWW-Authenticate header as `login.windows.net/12345`.  Normally MSAL would know that `login.microsoft.com` and `login.windows.net` are aliases, but when developers enable regional (WithAzureRegion), we are not allowed to make global calls, so we can't discover the aliases. So MSAL throws an exception. Hence we are trying to move away from `WithAuthority` at the request level and teach ppl to use `WithTenant` instead, because really only the tenant should change, not the environment part of the authority. We should hide `WithAuthority` at request level via EditorBrowsable attribute, maybe even soft deprecate it.

Both `WithTenantFromAuthority` or @MichelZ's proposal would solve this issue. I tend to agree with @pmaytak  and @jmprieur that `WithTenantFromAuthority` would be more discoverable.

Reply:
Ah nice, makes sense now, thanks!

Reply:
Like this? 
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/344fd591c26e9567107622dd18f2557739a62836/src/client/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs#L336

I'm struggling where to put tests for it :)

Reply:
For tests, AuthorityTests.cs maybe? There are already some tests for WithTenantId. I think the AcquireTokenXYZ methods should return back config instance, that you can check if authority was updated. It also seems like we don't have many specific tests for the current WithTenantId method, so maybe if you could add some for this too (which should be almost exactly the same as for WithTenantIdFromAuthority)

Reply:
Thanks @pmaytak. Have a look [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/de169bf00de3a6e0c8e2145e218f7383124376a2/tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs#L82) please for the new tests, thanks!